src/shared/utilities/collectionUtils.ts

```diff
@@ -160,3 +160,38 @@ export async function take<T>(sequence: AsyncIterable<T>, count: number): Promis
 
     return result
 }
+
+export interface getPaginatedAwsCallIterParams<TRequest, TResponse> {
+    awsCall: (request: TRequest) => Promise<TResponse>
+    nextTokenNames: {
+        request: keyof TRequest
+        response: keyof TResponse
+    }
+    request: TRequest
+}
+
+/**
+ * Generates an iterator representing a paginated AWS call from a request and an AWS SDK call
+ * Each next() call will make a new request with the previous request's nextToken.
+ * @param params Iterator params
+ */
+export async function* getPaginatedAwsCallIter<TRequest, TResponse>(
+    params: getPaginatedAwsCallIterParams<TRequest, TResponse>
+): AsyncIterator<TResponse> {
+    let nextToken: string | undefined = undefined
+
+    while (true) {
+        const response: TResponse = await params.awsCall({
+            ...params.request,
+            [params.nextTokenNames.request]: nextToken,
+        })
+        if (response[params.nextTokenNames.response]) {
+            nextToken = (response[params.nextTokenNames.response] as any) as string
```

Comment:
If you want to make this typesafe, you should be able to construct a subset of the original type that just contains string fields with a method simular to this.

https://stackoverflow.com/questions/50900533/how-can-i-extract-the-names-of-all-fields-of-a-specific-type-from-an-interface-i

Reply:
I don't think that'll work here since we're using a generic. I've tried that in the past with very mixed success.